luci2-ui-base/www/luci-ng/controller/system/upgrade.js

```diff
@@ -0,0 +1,145 @@
+L2.registerFactory('fileService', function() {
+    var files = [];
+    return files;
+});
+
+L2.registerDirective('fileModel', [ '$parse', 'fileService', function($parse, fileService) {
+	return {
+		restrict: 'A',
+		link: function($scope, element) {
+			element.bind('change', function() {
+				$scope.$apply(function(){
+					if (element[0].files != undefined) {
+						fileService.pop();
+						fileService.push(element[0].files[0]);
+					}
+				});
+			});
+		}
+	};
+}]);
+
+L2.registerController('SystemUpgradeController',['$modal', 'l2cgi', 'l2rpc', 'l2spin', 'fileService', 'gettext', function($modal, l2cgi, l2rpc, l2spin, fileService, gettext) {
+	var upgradeCtrl = this;
+
+	angular.extend(upgradeCtrl, {
+		getUploadFile: function(){
+			return fileService[0];
+		},
+
+		getSelectedFileName: function(log) {
+			var file=upgradeCtrl.getUploadFile();
+			if (file == undefined ) {
+				return gettext("No file selected");
+			}
+			else {
+				return file.name;
+			}
+		},
+
+		uploadFileSelected: function(){
+			if (upgradeCtrl.getUploadFile() != undefined ) {
+				return false;
+			}
+			else {
+				return true;
+			}
+		},
+
+		upgradeTestRpc: l2rpc.declare({
+			object: 'rpc-sys',
+			method: 'upgrade_test'
+		}),
+
+		upgradeStartRpc: l2rpc.declare({
+			object: 'rpc-sys',
+			method: 'upgrade_start',
+			params: [ 'keep' ]
+		}),
+
+		upgradeCleanRpc: l2rpc.declare({
+			object: 'rpc-sys',
+			method: 'upgrade_clean'
+		}),
+
+		reboot: l2rpc.declare({
+			object: 'luci2.system',
+			method: 'reboot'
+		}),
+
+		displayUpgradeSuccessCtrl: function ($scope, $modalInstance) {
+			var dialog = this;
+			$scope.data = upgradeCtrl.data;
+			return angular.extend(dialog, {
+				upgradeCtrl: upgradeCtrl,
+				upgrade: function() {
+					$modalInstance.dismiss();
+					upgradeCtrl.startUpgrade();
+				},
+				cancel: function() {
+					$modalInstance.dismiss();
+					upgradeCtrl.cleanUpgrade();
+				}
+			});
+		},
+
+		displayUpgradeFailedCtrl: function ($scope, $modalInstance) {
+			var dialog = this;
+			$scope.data = upgradeCtrl.data;
+			$modalInstance.opened.then(function () {
+				upgradeCtrl.cleanUpgrade();
+			});
+			return angular.extend(dialog, {
+				upgradeCtrl: upgradeCtrl,
+				confirm: function() {
+					$modalInstance.dismiss();
+				}
+			});
+		},
+
+		uploadUpgrade: function() {
+			var file = upgradeCtrl.getUploadFile();
+			l2spin.open();
+			l2cgi.uploadUpgrade(file).then(function(data) {
+				upgradeCtrl.data=data;
+				return upgradeCtrl.upgradeTestRpc();
+			}).then(function(data){
+				upgradeCtrl.data=data;
+				l2spin.close();
+				if ( data.code == 0 ){
```

Comment:
Don't use space within the parens but add one space after the closing paren:
`if (data.code == 0) {`